File: source/Game/Player.lua
class('Player').extends()

function Player:init(camera, world)
    self.camera = camera
    self.world = world
    self.velocity = Vector3(0, 0, 0)
    self.acceleration = 1.0 -- Impulse per frame of cranking
    self.friction = 0.95 -- Water drag
    self.lookSpeed = 0.05 -- Radians per frame
    self.wobbleSpeed = 0.002
    self.wobbleAmount = 0.02
end

function Player:update()
    -- Water Wobble (Roll)
    local time = playdate.getCurrentTimeMilliseconds()
    self.camera.rotation.z = math.sin(time * self.wobbleSpeed) * self.wobbleAmount
    
    -- D-Pad for Look (Pitch/Yaw)
    if playdate.buttonIsPressed(playdate.kButtonUp) then
        self.camera.rotation.x -= self.lookSpeed -- Look Up
    elseif playdate.buttonIsPressed(playdate.kButtonDown) then
        self.camera.rotation.x += self.lookSpeed -- Look Down
    end
    
    if playdate.buttonIsPressed(playdate.kButtonLeft) then
        self.camera.rotation.y -= self.lookSpeed -- Look Left
    elseif playdate.buttonIsPressed(playdate.kButtonRight) then
        self.camera.rotation.y += self.lookSpeed -- Look Right
    end
    
    -- Clamp Pitch to avoid flipping (optional, but good for FPS)
    local maxPitch = math.rad(89)
    if self.camera.rotation.x > maxPitch then self.camera.rotation.x = maxPitch end
    if self.camera.rotation.x < -maxPitch then self.camera.rotation.x = -maxPitch end
    
    -- Crank for Swimming (Throttle/Impulse)
    local crankChange = playdate.getCrankChange()
    
    if crankChange ~= 0 then
        -- Apply impulse based on crank speed and direction
        -- Positive crank (clockwise) -> Move Forward
        -- Negative crank (counter-clockwise) -> Move Backward
        
        local impulse = 0
        if crankChange > 0 then
            impulse = self.acceleration * (crankChange / 5) -- Scale by speed
        else
            impulse = self.acceleration * (crankChange / 5)
        end
        
        -- Get Forward Vector (Look Direction)
        local forward = self.camera:getForwardVector()
        
        -- Apply acceleration in look direction
        self.velocity = self.velocity:add(forward:mul(impulse))
    end
    
    -- Apply Friction
    self.velocity = self.velocity:mul(self.friction)
    
    -- Stop if very slow
    if self.velocity:magnitude() < 0.01 then
        self.velocity = Vector3(0, 0, 0)
    end
    
    -- Apply Velocity to Position
    if self.velocity:magnitude() > 0 then
        local newPos = self.camera.position:add(self.velocity)
        
        -- Collision with World Cube
        if self.world then
            local limit = self.world.size - 5
            
            if newPos.x > limit then 
                newPos.x = limit 
                self.velocity.x = 0 
            elseif newPos.x < -limit then 
                newPos.x = -limit 
                self.velocity.x = 0 
            end
            
            if newPos.y > limit then 
                newPos.y = limit 
                self.velocity.y = 0 
            elseif newPos.y < -limit then 
                newPos.y = -limit 
                self.velocity.y = 0 
            end
            
            if newPos.z > limit then 
                newPos.z = limit 
                self.velocity.z = 0 
            elseif newPos.z < -limit then 
                newPos.z = -limit 
                self.velocity.z = 0 
            end
        end
        
        self.camera.position = newPos
    end
end

File: source/Game/Player.c
#include "Player.h"

#include <math.h>

void player_init(Player* player, Camera* camera, World* world) {
    player->camera = camera;
    player->world = world;
    player->velocity = vector3_create(0.0f, 0.0f, 0.0f);
    player->acceleration = 1.0f;
    player->friction = 0.95f;
    player->look_speed = 0.05f;
    player->wobble_speed = 0.002f;
    player->wobble_amount = 0.02f;
}

void player_update(Player* player, PlaydateAPI* pd) {
    const uint32_t current_time_ms = pd->system->getCurrentTimeMilliseconds();
    player->camera->rotation.z = sinf((float)current_time_ms * player->wobble_speed) * player->wobble_amount;

    PDButtons current;
    PDButtons pushed;
    PDButtons released;
    pd->system->getButtonState(&current, &pushed, &released);

    if ((current & kButtonUp) != 0) {
        player->camera->rotation.x -= player->look_speed;
    } else if ((current & kButtonDown) != 0) {
        player->camera->rotation.x += player->look_speed;
    }

    if ((current & kButtonLeft) != 0) {
        player->camera->rotation.y -= player->look_speed;
    } else if ((current & kButtonRight) != 0) {
        player->camera->rotation.y += player->look_speed;
    }

    const float max_pitch = 1.553343f;
    if (player->camera->rotation.x > max_pitch) {
        player->camera->rotation.x = max_pitch;
    }
    if (player->camera->rotation.x < -max_pitch) {
        player->camera->rotation.x = -max_pitch;
    }

    const float crank_change = pd->system->getCrankChange();
    if (crank_change != 0.0f) {
        const float impulse = player->acceleration * (crank_change / 5.0f);
        const Vector3 forward = camera_get_forward_vector(player->camera);
        player->velocity = vector3_add(player->velocity, vector3_mul(forward, impulse));
    }

    player->velocity = vector3_mul(player->velocity, player->friction);

    if (vector3_magnitude(player->velocity) < 0.01f) {
        player->velocity = vector3_create(0.0f, 0.0f, 0.0f);
    }

    if (vector3_magnitude(player->velocity) > 0.0f) {
        Vector3 new_position = vector3_add(player->camera->position, player->velocity);

        if (player->world != NULL) {
            const float limit = (float)player->world->size - 5.0f;

            if (new_position.x > limit) {
                new_position.x = limit;
                player->velocity.x = 0.0f;
            } else if (new_position.x < -limit) {
                new_position.x = -limit;
                player->velocity.x = 0.0f;
            }

            if (new_position.y > limit) {
                new_position.y = limit;
                player->velocity.y = 0.0f;
            } else if (new_position.y < -limit) {
                new_position.y = -limit;
                player->velocity.y = 0.0f;
            }

            if (new_position.z > limit) {
                new_position.z = limit;
                player->velocity.z = 0.0f;
            } else if (new_position.z < -limit) {
                new_position.z = -limit;
                player->velocity.z = 0.0f;
            }
        }

        player->camera->position = new_position;
    }
}
